sources/TileDB.CSharp/Query.cs

```diff
@@ -1058,66 +1350,79 @@ private DataType GetDataType(string name)
         /// </summary>
         private void FreeAllBufferHandles()
         {
-            foreach (var bh in _dataBufferHandles)
-            {
-                bh.Value.Free();
-            }
+            DisposeValuesAndClear(_dataBufferHandles);
+            DisposeValuesAndClear(_offsetsBufferHandles);
+            DisposeValuesAndClear(_validityBufferHandles);
 
-            foreach (var bh in _offsetsBufferHandles)
+            static void DisposeValuesAndClear(Dictionary<string, BufferHandle> handles)
             {
-                bh.Value.Free();
+                foreach (var bh in handles)
+                {
+                    bh.Value.Dispose();
+                }
+                handles.Clear();
             }
+        }
 
-            foreach (var bh in _validityBufferHandles)
+        private static void AddOrReplaceBufferHandle(Dictionary<string, BufferHandle> dict, string name, BufferHandle handle)
+        {
+            if (dict.TryGetValue(name, out var existingHandle))
             {
-                bh.Value.Free();
+                existingHandle.Dispose();
             }
+            dict[name] = handle;
         }
 
-        /// <summary>
-        /// Add data buffer handle.
-        /// </summary>
-        /// <param name="name"></param>
-        /// <param name="handle"></param>
-        /// <param name="size"></param>
-        private void AddDataBufferHandle(string name, GCHandle handle, ulong size)
+        private sealed class BufferHandle : IDisposable
         {
-            if (_dataBufferHandles.ContainsKey(name))
+            private MemoryHandle DataHandle;
+            public ulong* SizePointer { get; private set; }
+
+            public void* DataPointer => DataHandle.Pointer;
+
+            public ulong Size
             {
-                _dataBufferHandles[name].Free();
+                get
+                {
+                    Debug.Assert(SizePointer is not null);
+                    return *SizePointer;
+                }
+                set
+                {
+                    Debug.Assert(SizePointer is not null);
+                    *SizePointer = value;
+                }
             }
-            _dataBufferHandles[name] = new BufferHandle(handle, size);
-        }
 
-        /// <summary>
-        /// Add offset buffer handle.
-        /// </summary>
-        /// <param name="name"></param>
-        /// <param name="handle"></param>
-        /// <param name="size"></param>
-        private void AddOffsetsBufferHandle(string name, GCHandle handle, ulong size)
-        {
-            if (_offsetsBufferHandles.ContainsKey(name))
+            public BufferHandle(ref MemoryHandle handle, ulong size)
```

Comment:
We pass the `MemoryHandle` by reference to ensure it exists in only one place in memory and avoid disposing it twice. Disposing `default(MemoryHandle)` does nothing.